File: src/_01_CreationalPattern/_01_04_Builder/java/after/DefaultTourPlanBuilder.java
package _01_CreationalPattern._01_04_Builder.java.after;

import java.time.LocalDate;
import java.util.ArrayList;

public class DefaultTourPlanBuilder implements TourPlanBuilder {

  private TourPlan tourPlan;

  @Override
  public TourPlanBuilder initTourPlan() {
    tourPlan = new TourPlan();
    return this;
  }

  private void checkTourPlan() {
    if(tourPlan == null) throw new Error("initTourPlan() 을 먼저 호출해 주세요");
  }
  
  @Override
  public TourPlanBuilder nightsAndDays(int nights, int days) {
    checkTourPlan();
    tourPlan.setNights(nights);
    tourPlan.setDays(days);
    return this;
  }

  @Override
  public TourPlanBuilder title(String title) {
    checkTourPlan();
    tourPlan.setTitle(title);
    return this;
  }

  @Override
  public TourPlanBuilder startDate(LocalDate startDate) {
    checkTourPlan();
    tourPlan.setStartDate(startDate);
    return this;
  }

  @Override
  public TourPlanBuilder whereToStay(String whereToStay) {
    checkTourPlan();
    tourPlan.setWhereToStay(whereToStay);
    return this;
  }

  @Override
  public TourPlanBuilder addPlan(int day, String plan) {
    checkTourPlan();
    
    if(tourPlan.getPlans() == null) tourPlan.setPlans(new ArrayList<>());
    
    tourPlan.addPlan(day, plan);
    return this;
  }

  @Override
  public TourPlan build() {
    TourPlan resultTourPlan = tourPlan;
    tourPlan = null;
    return resultTourPlan;
  }
  
}

File: src/_01_CreationalPattern/_01_04_Builder/DefaultTourPlanBuilder.js
import TourPlan from "./TourPlan.js";
import TourPlanBuilder from "./TourPlanBuilder.js";

export default class DefaultTourPlanBuilder extends TourPlanBuilder {
  /** @type { TourPlan } */
  #tourPlan;
  
  /** 
   * @override 
   * @returns { DefaultTourPlanBuilder }
   */
  initTourPlan() {
    this.#tourPlan = new TourPlan();
    return this;
  }

  #checkTourPlan() {
    if(!this.#tourPlan) {
      throw new Error("initTourPlan() 을 먼저 호출해 주세요");
    }
  }

  /**
   * @override
   * @param { string } title
   * @returns { DefaultTourPlanBuilder }
   */
  title(title) {
    this.#checkTourPlan();
    this.#tourPlan.setTitle(title);
    return this;
  }

  /**
   * @override
   * @param { number } nights
   * @param { number } days
   * @returns { DefaultTourPlanBuilder }
   */
  nightsAndDays(nights, days) {
    this.#checkTourPlan();
    this.#tourPlan.setNights(nights);
    this.#tourPlan.setDays(days);
    return this;
  }

  /**
   * @override
   * @param { Date } startDate
   * @return { DefaultTourPlanBuilder }
   */
  startDate(startDate) {
    this.#checkTourPlan();
    this.#tourPlan.setStartDate(startDate);
    return this;
  }

  /**
   * @override
   * @param { string } whereToStay
   * @returns { DefaultTourPlanBuilder }
   */
  whereToStay(whereToStay) {
    this.#checkTourPlan();
    this.#tourPlan.setWhereToStay(whereToStay);
    return this;
  }

  /**
   * @override
   * @param { number } day
   * @param { string } plan
   * @returns { DefaultTourPlanBuilder }
   */
  addPlan(day, plan) {
    this.#checkTourPlan();
    this.#tourPlan.addPlan(day, plan);
    return this;
  }

  /**
   * @override
   * @returns { TourPlan }
   */
  build() {
    this.#checkTourPlan();
    const resultTourPlan = this.#tourPlan;
    this.#tourPlan = undefined;
    return resultTourPlan;
  }
}